### ISSUE TITLE
ExcelAsyncUtil.QueueAsMacro in a background thread and saving a file on a network drive interfering with VBA event processing and causing 0x800AC472 busy errors

### ISSUE THREAD

**Issue by gmichaud** [Excel-DNA/ExcelDna#353]
We are seeing a really weird issue in production with one of our customers, and after a very long troubleshooting session we have narrowed down the issue to two things:

- The user is saving a file to a network drive over a VPN, causing Excel to show a modal "Saving..." dialog when saving the file
- ExcelAsyncUtil.QueueAsMacro is being invoked in a background thread on a regular basis

I have created a minimal example to replicate this issue. The Excel file has a VBA macro that shows a message box in the AfterSave event, and we are also listening to the WorkbookAfterSave event in the .NET add-in.

When saving the same file to a local drive, everything works as expected. The message box in the VBA macro shows up, and the code that we have in the add-in WorkbookAfterSave event handler executes successfully. When saving the same file to the network drive, we see two issues:

- The VBA AfterSave event is not raised at all
- Our add-in WorkbookAfterSave event is raised, but we are not able to inspect all the properties of the workbook; a COMExeception with HRESULT 0x800AC472 ("busy") is raised

I have recorded a short YouTube video to illustrate the issue: https://www.youtube.com/watch?v=OXokOjgNnXw

In the sample add-in code we're not doing anything with the Excel object model but in our production app we are retrieving the calculation state of the application.

The add-in code:

    public class AddIn : IExcelAddIn
    {
        public void AutoClose()
        {
        }

        public void AutoOpen()
        {
            var excelApp = (Microsoft.Office.Interop.Excel.Application) ExcelDnaUtil.Application;
            excelApp.WorkbookAfterSave += ExcelApp_WorkbookAfterSave;

            Task.Run(() =>
            {
                while(true)
                {
                    System.Threading.Thread.Sleep(2000);
                    ExcelAsyncUtil.QueueAsMacro(() =>
                    {
                        System.Diagnostics.Debug.WriteLine(DateTime.Now.ToString() + " Check calculation state");
                    });
                }
            });

            System.Diagnostics.Debug.WriteLine("AutoOpen() initialization complete");
        }

        private void ExcelApp_WorkbookAfterSave(Workbook Wb, bool Success)
        {
            //If file save triggers the "Saving..." modal dialog (seems to happen when saving on a network drive over a VPN or slow connection?)
            //then accessing some of the properties of the workbook will return a COMException with  HRESULT 0x800AC472
            try
            {
                var sheet = (Worksheet)Wb.Sheets[1];
                int usedRangeRowsCount = sheet.UsedRange.Rows.Count;
                System.Windows.Forms.MessageBox.Show($"In Addin WorkbookAfterSave. Sheet used range has {usedRangeRowsCount} rows.", "", System.Windows.Forms.MessageBoxButtons.OK);
            }
            catch(Exception ex)
            {
                System.Windows.Forms.MessageBox.Show($"In AddIn WorkbookAfterSave -- failed to retrieve row count. Exception: {ex.Message}.", "", System.Windows.Forms.MessageBoxButtons.OK);
            }
        }
    }

**Comment by gmichaud** [comment]
I just thought of a way to work around the issue -- we can pause all our background tasks in BeforeSave and only resume them once save is complete.

**Comment by govert** [comment]
It's a strange problem. I imagine the `QueueAsMacro` won't run during the save (when the Save dialog is shown) but things are too quick to see that. But then the moment the save is finished, both the QueueAsMacro and the AfterSave event runs (both on the main thread). So I think this is an Excel bug where the internal guard against running COM code is somehow not getting set or cleared at the right time. I'm not sure how you'd turn this into a bug report for Microsoft, though.

Your workaround sound like a reasonable plan.

**Comment by gmichaud** [comment]
Thanks for your feedback @govert. Upon further review we also found another weird issue that we had ran into recently that happen when opening files from WebDav; there's a modal dialog that shows up and that also generates strange issues with any code that we try to run through QueueAsMacro() from app events. I don't want to rely on a bug fix from Microsoft, so we will work around the issue ourselves as described above.
